File: 5_mergesort/python/TopDownMergeSort.py
class TopDownMergeSort:

    @staticmethod
    def sort(a):
        _sort(a, 0, len(a) - 1)

    @staticmethod
    def _sort(a, lo, hi):
        if (hi <= lo):
            return
        
        mid = lo + (hi - lo) // 2
        _sort(a, lo, mid)
        _sort(a, mid + 1, hi)
        merge(a, lo, mid, hi)

    @staticmethod
    def merge(a, lo, mid, hi):
        i = lo
        j = mid
        aux = list()

        for k in range(hi + 1):
            aux[k] = a[k]

        for k in range(hi + 1):
            if i > mid:
                a[k] = aux[j]
                j += 1
            elif j > hi:
                a[k] = aux[i]
                i += 1
            elif less(aux[j], aux[i]):
                a[k] = aux[j]
                j += 1
            else:
                a[k] = aux[i]
                i

    @staticmethod
    def less(v, w):
        return v < w

    @staticmethod
    def show(a):
        print(a)

    

File: 5_mergesort/java/TopDownMergeSort.java
public class TopDownMergeSort {
    private static Comparable[] aux;

    public static void sort(Comparable[] a)
    {
        aux = new Comparable[a.length];
        sort(a, 0, a.length - 1);
    }
    
    private static void sort(Comparable a[], int lo, int hi)
    {
        if (hi <= lo) return;
        int mid = lo + (hi - lo) / 2;
        sort(a, lo, mid);
        sort(a, mid + 1, hi);
        merge(a, lo, mid, hi);
    }

    private static void merge(Comparable a[], int lo, int mid, int hi)
    {
        int i = lo, j = mid;

        for (int k = lo; k <= hi; k++)
        {
            aux[k] = a[k];
        }

        for (int k = lo; k <= hi; k++)
        {
            if (i > mid)                    a[k] = aux[j++];  // left side is exhausted
            else if (j > hi)                a[k] = aux[i++];  // right side is exhaused
            else if (less(aux[j], aux[i]))  a[k] = aux[j++];  // right key is less
            else                            a[k] = aux[i++];  // left key is less or equal
        }
    }

    private static boolean less(Comparable v, Comparable w)
    {
        return v.compareTo(w) < 0;
    }

    public static void show(String[] a)
    {
        for (int i = 0; i < a.length; i++)
        {
            StdOut.print(a[i] + " ");
            StdOut.println();
        }
    }
}